### ISSUE TITLE
Feature Request: .NET MAUI/mobile support 🌺

### ISSUE THREAD

**Issue by maddymontaquila** [dotnet/aspire#4684]
cc @BretJohnson @bradygaster 

It would be great to launch .NET MAUI apps from the AppHost file, see them on the dashboard, and auto-hook-up all the endpoints and such. This *sort-of* works but I would like to formalize it and make sure it works well and aligned with how aspire actually does its thing

**Requirements**
* Can add/reference .NET MAUI app in the AppHost file
* Ability to choose which MAUI target to launch: via apphost? something in VS or VSCode tooling? Something in the dashboard perhaps would be cool - "Start MAUI app" > popup to choose device???
* Endpoint configuration - ie, the Android emulator/iOS sim/device/local Windows or mac app should use some form of devtunnel etc to talk to the APIs and other containers spun up
* Support for OTEL? I have no idea if this would "just work" or not
* NON GOAL - deployment is unnecessary unless Aspire gets a really great CI/CD story down the line. Mobile apps aren't really... deployed lol. But they use plenty of deployed things!

Bret had taken a stab at a proof of concept of this here https://github.com/BretJohnson/aspire-mobile - I think we need to change how we're doing launch in this though (currently, you ctrl+f5 the app host then start the maui app separately... ideally we do something better than that!)

For reference, this is how you currently choose a debug target/TFM (maui uses multi-targeting) in VS and VS Code when you debug a MAUI app. you often change between different targets, so we need a way for people to do this in aspire. 

https://github.com/dotnet/aspire/assets/12660687/2cee4c1b-4daa-4d41-bcc7-3cc21120d11b

https://github.com/dotnet/aspire/assets/12660687/0c719085-1b44-4e00-bf47-56446a0cbf85

**Comment by maddymontaquila** [comment]
Could possibly be related to https://github.com/dotnet/aspire/issues/813

**Comment by BretJohnson** [comment]
FYI - MAUI work to support Aspire is being tracked here: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2000198/. But let's use this issue to track the Aspire support that's needed as part of that.

**Comment by maddymontaquila** [comment]
we could put it here, same thing as web browser just like... "Device Target" or something? 
![image](https://github.com/dotnet/aspire/assets/12660687/0e48dc61-3d79-43e1-aba7-cc6cb3c20398)

**Comment by danmoseley** [comment]
Eshop is aspirified and has a mobile app - I assume it would benefit?

**Comment by BretJohnson** [comment]
> we could put it here, same thing as web browser just like... "Device Target" or something? ![image](https://private-user-images.githubusercontent.com/12660687/343462702-0e48dc61-3d79-43e1-aba7-cc6cb3c20398.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0NTMxOTcsIm5iZiI6MTcxOTQ1Mjg5NywicGF0aCI6Ii8xMjY2MDY4Ny8zNDM0NjI3MDItMGU0OGRjNjEtM2Q3OS00M2UxLWFiYTctY2M2Y2IzYzIwMzk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDAxNDgxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3NGMyMzYxYjYzYjgxNjU4OTk1YWI5OTFlNDc2M2FjZWM2YjczYmM3MzdhNWNlYjljMmI4MTc3YzM0YTVkNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2diDnczw4cXjyy8EhZBLB7Pv4VKId1v-EN7FxkUqYHI)

Yes, potentially. FYI, when we discussed this back in January @DamianEdwards said there are other scenarios, beyond MAUI, where they'd like to show some kind of launch UI in the future. I asked for examples and he said this:

![image](https://github.com/dotnet/aspire/assets/245892/0e63292f-3dd0-4a4d-93ca-10f35c51a763)

**Comment by bijington** [comment]
To give my input on the possible options mentioned in https://github.com/dotnet/aspire/issues/4684#issuecomment-2192668061

While I would probably prefer option 2 initially because it fits more with the expected workflow of a .NET MAUI app developer I think option 1 would likely be a better fit as it gives a consistent approach to the current Aspire model as you said @BretJohnson

**Comment by maddymontaquila** [comment]
> Eshop is aspirified and has a mobile app - I assume it would benefit?

You know it!!! This was the main reason we didnt "finish" wiring up the eshop mobile client - no way to launch them both without it being gross.

**Comment by maddymontaquila** [comment]
> To give my input on the possible options mentioned in [#4684 (comment)](https://github.com/dotnet/aspire/issues/4684#issuecomment-2192668061)
> 
> While I would probably prefer option 2 initially because it fits more with the expected workflow of a .NET MAUI app developer I think option 1 would likely be a better fit as it gives a consistent approach to the current Aspire model as you said @BretJohnson


I'm with you here Shaun. I think 1 is the "right" thing to do from the Aspire perspective. Maybe we call that the MVP and listen to see if that really confuses people or it works okay. I don't think we'd need to separate out by TFM either - just a list of available debug targets we find.

**Comment by BretJohnson** [comment]
> > To give my input on the possible options mentioned in [#4684 (comment)](https://github.com/dotnet/aspire/issues/4684#issuecomment-2192668061)
> > While I would probably prefer option 2 initially because it fits more with the expected workflow of a .NET MAUI app developer I think option 1 would likely be a better fit as it gives a consistent approach to the current Aspire model as you said @BretJohnson
> 
> I'm with you here Shaun. I think 1 is the "right" thing to do from the Aspire perspective. Maybe we call that the MVP and listen to see if that really confuses people or it works okay. I don't think we'd need to separate out by TFM either - just a list of available debug targets we find.

Though worth noting - I think that 1 is the more expensive option technically.

**Comment by bijington** [comment]
@BretJohnson When you say it is more expensive (not that I am footing the bill) but do you have any thoughts on how much more expensive over option 2? Could there be an option that starts with option 2 (or possibly a new option) but puts infrastructure in place for option 1?

**Comment by BretJohnson** [comment]
@bijington - In terms of cost, we still need to design (design both UX and implementation) and cost it. And I think we should design and cost. I believe that @bradygaster will be looking at the experience design. Dev cost wise, I'd say probably at least a couple man months, but I'm just swaging and we need to dig in - which we should.

My main concern here is just to make sure we have a good sense of what customers (like yourself) actually want to do - what workflows want to support well. And use that knowledge to drive the spec / priorities. It'd be a shame to devote a lot of effort to supporting option 1, and only option 1, if option 2 is what folks actually would prefer most often in practice. Or maybe we should support both. For instance, most of the time when debugging the MAUI app would you also want to debug all the services? Most of the time when restarting the MAUI app would you want to restart all the services?

**Comment by maddymontaquila** [comment]
Ok so the current POC - 
- Everything is driven via the AppHost. You start debugging the apphost, and that will start the MAUI app along with all the other things. **Let's cost what it'll be for that to happen** if we use the launch UI @DamianEdwards had shared earlier to select a target device. (Also we can do something in that popout menu that says - to manage MAUI devices change your startup proj)
- Reference the MAUI proj in AppHost either with AddMauiApp or AddClientApp/AddNativeApp - leaning towards AddMauiApp so we don't accidentally pigeonhole ourselves into having supporting everything else right away
- Something something devtunnels @sayedihashimi is going to sort out LOL (related to #813 )
- The apphost/DCP should probably try to start the maui app ABSOLUTELY LAST so that it doesnt slow down the other builds - maui usually takes the longest. Maybe implicitly we use something like WaitOn() (or whatever we do for #921 ) when that gets implemented but user could override
- this will be icky until we implement something to help w #3095 since youll have to restart everything including the maui app every time and as we all know mobile b slow

cc @bradygaster @BretJohnson

**Comment by maddymontaquila** [comment]
@francoistanguay yeah absolutely!!! ideally aspire apphost knows nothing about the projects its referencing, so im sure we could extend it to work somehow with Uno, or at least hook you guys up in your extension

**Comment by maddymontaquila** [comment]
Update for the folks following along - MAUI platform team (@emaf ) is gonna look at #3043 when they get a chance because that is the first thing we hit with apphost blocking us. There will also be some Android SDK changes needed to pass launch variables and such.

**Comment by maddymontaquila** [comment]
also related https://github.com/dotnet/aspire/issues/876

**Comment by Robbert-jansen** [comment]
Omg, this would save my life

**Comment by bricefriha** [comment]
I would just like to say that this would make more sense (cents) than a changing machine. (This sentence works better when being spoken).

Either way, that would be a game-changer! 
Pretty please!?

**Comment by burtonrodman** [comment]
in Visual Studio, couldn't the launch experience be handled with a Multiple startup projects profile?  This also solves the target issue.  This is working on my machine...

![Image](https://github.com/user-attachments/assets/2d070df3-2dde-4579-a645-3afc8c1f2d90)

The only other issue I really need to solve at this point is creating a devtunnel and passing the hostname to the Maui app -- could we break this down to a smaller Issue and solve just this?

**Comment by moljac** [comment]
@burtonrodman 

> in Visual Studio, couldn't the launch experience be handled with a Multiple startup projects profile? This also solves the target issue. This is working on my machine...

I was launching everything from CLI by running AppHost. I stopped because I was blocked by endpoint issue:

https://github.com/dotnet/aspire/issues/4868

... and I was not aware that there is workaround. I will resurrect my attempts as soon as I will be able. Namely I want to rewrite APIs for emulators/simulators with CliWrap I have feeling I might be faster with it.


Few videos with clients and backends for MAUI and even for Python (no dockerizing)

*   https://www.youtube.com/watch?v=fx1XOUFdU8k

*   https://www.youtube.com/watch?v=nnFf58Ff1jA

*   https://www.youtube.com/watch?v=zkvAaSTPs0c

*   https://www.youtube.com/watch?v=oDSW0t_x9MM

**Comment by maddymontaquila** [comment]
Hey this works now!
